File: source/XYO/Fabricare/Library.js
// Created by Grigore Stefan <g_stefan@yahoo.com>
// Public domain (Unlicense) <http://unlicense.org>
// SPDX-FileCopyrightText: 2021-2023 Grigore Stefan <g_stefan@yahoo.com>
// SPDX-License-Identifier: Unlicense

Script.requireExtension("Console");
Script.requireExtension("Application");
Script.requireExtension("Buffer");
Script.requireExtension("DateTime");
Script.requireExtension("JSON");
Script.requireExtension("Shell");
Script.requireExtension("ShellFind");
Script.requireExtension("Math");
Script.requireExtension("ProcessInteractive");
Script.requireExtension("SHA512");
Script.requireExtension("Thread");
Script.requireExtension("CSV");

// ---

Script.resetIncludePath();

// ---

Fabricare.configFile = "";
Fabricare.userConfigFile = "";
Fabricare.workspaceFile = "";
Fabricare.action = Application.getArgument(0, "default");

// ---

Config = {};

Fabricare.loadConfig = function() {
	var cwd = Shell.getcwd();
	if (!Script.isNil(cwd)) {
		cwd += "/";
	};

	Fabricare.configFile = Application.getFlagValue("config", cwd + ".fabricare.json");

	var json = Shell.fileGetContents(Fabricare.configFile);
	if (!Script.isNil(json)) {
		Config = JSON.decode(json);
		if (!Script.isNil(global.Config)) {
			return true;
		};
	};

	global.Config = {};
	return false;
};

Fabricare.saveConfig = function() {
	var cwd = Shell.getcwd();
	if (!Script.isNil(cwd)) {
		cwd += "/";
	};

	Fabricare.configFile = Application.getFlagValue("config", cwd + ".fabricare.json");

	return Shell.filePutContents(Fabricare.configFile, JSON.encodeWithIndentation(Config));
};

// ---

UserConfig = {};

Fabricare.loadUserConfig = function() {
	var pathHome = null;
	if (OS.isWindows()) {
		if (OS.isMinGW()) {
			pathHome = Shell.getenv("HOME");
			if (Script.isNil(pathHome)) {
				pathHome = Shell.getenv("HOMEDRIVE") + Shell.getenv("HOMEPATH");
			};
		} else {
			pathHome = Shell.getenv("HOMEDRIVE") + Shell.getenv("HOMEPATH");
		};
	};
	if (OS.isLinux()) {
		pathHome = Shell.getenv("HOME");
	};
	if (Script.isNil(pathHome)) {
		Console.writeLn("* Error: Unable to determine user home path!");
		Script.exit(1);
	};

	Fabricare.userConfigFile = Application.getFlagValue("user-config", pathHome + ".fabricare.json");

	if (!Shell.fileExists(Fabricare.userConfigFile)) {
		global.UserConfig = {};
		return false;
	};

	var json = Shell.fileGetContents(Fabricare.userConfigFile);
	if (!Script.isNil(json)) {
		UserConfig = JSON.decode(json);
		if (!Script.isNil(global.UserConfig)) {
			return true;
		};
	};

	global.UserConfig = {};
	return false;
};

Fabricare.saveUserConfig = function() {
	var pathHome = null;
	if (OS.isWindows()) {
		if (OS.isMinGW()) {
			pathHome = Shell.getenv("HOME");
			if (Script.isNil(pathHome)) {
				pathHome = Shell.getenv("HOMEDRIVE") + Shell.getenv("HOMEPATH");
			};
		} else {
			pathHome = Shell.getenv("HOMEDRIVE") + Shell.getenv("HOMEPATH");
		};
	};
	if (OS.isLinux()) {
		pathHome = Shell.getenv("HOME");
	};
	if (Script.isNil(pathHome)) {
		Console.writeLn("* Error: Unable to determine user home path!");
		Script.exit(1);
	};

	Fabricare.userConfigFile = Application.getFlagValue("user-config", pathHome + ".fabricare.json");

	return Shell.filePutContents(Fabricare.userConfigFile, JSON.encodeWithIndentation(UserConfig));
};

// ---

Workspace = {};

// ---

Platform = {};
Platform.name = "unknwon";
Platform.machine = "unknwon";
Platform.osName = "unknwon";
Platform.osType = "unknwon";

// ---

Solution = {};
Solution.name = "unknwon";
Solution["SPDX-License-Identifier"] = "LicenseRef-Unknwon";
Solution.type = null;
Solution.projects = [];

// ---

Fabricare.include = function(file) {	
	var files = [
		// Local
		Shell.getcwd() + "/fabricare/solution/" + Solution.type + "." + file + ".js",
		Shell.getcwd() + "/fabricare/" + file + ".js",
		// Application
		Application.getPathExecutable() + "/fabricare/solution/" + Solution.type + "." + file + ".js",
		Application.getPathExecutable() + "/fabricare/" + file + ".js"
	];
	for (var index=0;index<files.length; ++index) {
		if (Shell.fileExists(files[index])) {			
			Script.include(files[index]);
			return true;
		};
	};
	files = [
		// Internal
		"fabricare://solution/" + Solution.type + "." + file + ".js",
		"fabricare://" + file + ".js"
	];
	for (var index=0;index<files.length; ++index) {
		if (Script.hasIncludeSource(files[index])) {			
			Script.include(files[index]);
			return true;
		};
	};
	return false;
};

// ---

fileToCS = Internal.fileToCS;
fileToRC = Internal.fileToRC;
htmlToRC = Internal.htmlToRC;
fileToJS = Internal.fileToJS;

xyoCC = function() {
	arguments.push("--platform=" + Platform.name);
	return Internal.xyoCC.apply(null, arguments);
};

xyoVersion = Internal.xyoVersion;

// ---

Fabricare.loadConfig();
Fabricare.loadUserConfig();

// ---
Fabricare.isDebug = function() {
	if (Application.hasFlag("debug")) {
		return true;
	};
	if (Shell.hasEnv("XYO_COMPILE_DEBUG")) {
		return true
	};
	return false;
};
Fabricare.isRelease = function() {
	return !Fabricare.isDebug();
};
// ---
Fabricare.loadWorkspace = function() {
	var cwd = Shell.getcwd();
	if (!Script.isNil(cwd)) {
		cwd += "/";
	};

	Fabricare.workspaceFile = Application.getFlagValue("workspace", cwd + "fabricare.json");

	var json = Shell.fileGetContents(Fabricare.workspaceFile);
	if (!Script.isNil(json)) {
		jsonWorkspace = JSON.decode(json);
		if (!Script.isNil(jsonWorkspace)) {
			for (var property in jsonWorkspace) {
				Workspace[property] = jsonWorkspace[property];
			};
		};
	};
};
Fabricare.loadWorkspace();
if (!Script.isNil(Workspace.solution)) {

	for (var property in Workspace.solution) {
		Solution[property] = Workspace.solution[property];
	};
};

if (Script.isNil(Solution.type)) {
	Solution.type = "generic";
};

// ---

Fabricare.processSolution = function() {	
	return Fabricare.include("solution/" + Solution.type + ".solution");
};

Fabricare.processWorkspace = function() {
	var runWorkspace=Shell.getcwd() + "/fabricare/workspace.js";
	if (Shell.fileExists(runWorkspace)) {
		return Script.include(runWorkspace);
	};
	return Fabricare.processSolution();
};

// ---

(function() {

if (Application.getFlagValue("run-script")) {
	return;
};

// ---

Platform.name = Application.getFlagValue("platform");
if (!Platform.name) {
	Fabricare.include("platform/detect");
};
if (!Fabricare.include("platform/" + Platform.name)) {
	Console.writeLn("Error: Platform " + Platform.name + " not found!");
	Script.exit(1);
};

// ---
})();

File: source/XYO/Fabricare/Library.cpp
// Fabricare
// Copyright (c) 2021-2025 Grigore Stefan <g_stefan@yahoo.com>
// MIT License (MIT) <http://opensource.org/licenses/MIT>
// SPDX-FileCopyrightText: 2021-2025 Grigore Stefan <g_stefan@yahoo.com>
// SPDX-License-Identifier: MIT

#include <XYO/Fabricare/Library.hpp>
#include <XYO/QuantumScript.Extension/Console.hpp>
#include <XYO/QuantumScript.Extension/Buffer.hpp>
#include <XYO/QuantumScript.Extension/DateTime.hpp>
#include <XYO/QuantumScript.Extension/JSON.hpp>
#include <XYO/QuantumScript.Extension/Application.hpp>
#include <XYO/QuantumScript.Extension/Shell.hpp>
#include <XYO/QuantumScript.Extension/ShellFind.hpp>
#include <XYO/QuantumScript.Extension/Math.hpp>
#include <XYO/QuantumScript.Extension/ProcessInteractive.hpp>
#include <XYO/QuantumScript.Extension/SHA512.hpp>
#include <XYO/QuantumScript.Extension/Thread.hpp>
#include <XYO/QuantumScript.Extension/Task.hpp>
#include <XYO/QuantumScript.Extension/Job.hpp>
#include <XYO/QuantumScript.Extension/Make.hpp>
#include <XYO/QuantumScript.Extension/CSV.hpp>
#include <XYO/QuantumScript.Extension/URL.hpp>
#include <XYO/FileToCS.Application.hpp>
#include <XYO/FileToRC.Application.hpp>
#include <XYO/HTMLToRC.Application.hpp>
#include <XYO/CPPCompilerCommandDriver.Application.hpp>
#include <XYO/Version.Application.hpp>
#include <XYO/Fabricare/Version.hpp>

namespace XYO::Fabricare {

	template <typename TApplication>
	TPointer<Variable> internalCall(const char *cmd0, VariableArray *arguments) {
		int k, cmdN;
		char **cmdS;
		TDynamicArray<String> cmdV;
		TPointer<Variable> retV;

		retV = VariableNumber::newVariable(0);
		cmdN = arguments->value->length() + 1;
		cmdS = new char *[cmdN];
		cmdV[0] = cmd0;
		for (k = 1; k < cmdN; ++k) {
			cmdV[k] = (arguments->index(k - 1))->toString();
		};
		for (k = 0; k < cmdN; ++k) {
			cmdS[k] = const_cast<char *>(cmdV[k].value());
		};
		try {
			TApplication application;
			((VariableNumber *)retV.value())->value = (Number)(application.main(cmdN, cmdS));
		} catch (...) {
			((VariableNumber *)retV.value())->value = -1;
			delete cmdS;
			throw;
		};
		delete cmdS;
		return retV;
	};

	TPointer<Variable> internalFileToCS(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- internal-file-to-cs\n");
#endif
		return internalCall<FileToCS::Application::Application>("file-to-cs", arguments);
	};

	TPointer<Variable> internalFileToRC(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- internal-file-to-rc\n");
#endif
		return internalCall<FileToRC::Application::Application>("file-to-rc", arguments);
	};

	TPointer<Variable> internalHTMLToRC(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- internal-html-to-rc\n");
#endif
		return internalCall<HTMLToRC::Application::Application>("html-to-rc", arguments);
	};

	TPointer<Variable> internalXYOCC(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- internal-xyo-cc\n");
#endif
		return internalCall<CPPCompilerCommandDriver::Application::Application>("xyo-cc", arguments);
	};

	TPointer<Variable> internalXYOVersion(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- internal-xyo-version\n");
#endif
		return internalCall<::XYO::Version::Application::Application>("xyo-cc", arguments);
	};

	TPointer<Variable> xyoIsConfigDefined(VariableFunction *function, Variable *this_, VariableArray *arguments) {
		String isDefined((arguments->index(0))->toString());
// -- Operating System
#ifdef XYO_PLATFORM_OS_WINDOWS
		if (isDefined == "XYO_PLATFORM_OS_WINDOWS") {
			return VariableBoolean::newVariable(true);
		};
#endif
#ifdef XYO_PLATFORM_OS_LINUX
		if (isDefined == "XYO_PLATFORM_OS_LINUX") {
			return VariableBoolean::newVariable(true);
		};
#endif
#ifdef XYO_PLATFORM_OS_MINGW
		if (isDefined == "XYO_PLATFORM_OS_MINGW") {
			return VariableBoolean::newVariable(true);
		};
#endif

// -- Compiler
#ifdef XYO_PLATFORM_COMPILER_MSVC
		if (isDefined == "XYO_PLATFORM_COMPILER_MSVC") {
			return VariableBoolean::newVariable(true);
		};
#endif
#ifdef XYO_PLATFORM_COMPILER_GCC
		if (isDefined == "XYO_PLATFORM_COMPILER_GCC") {
			return VariableBoolean::newVariable(true);
		};
#endif

// -- Application
#ifdef XYO_PLATFORM_64BIT
		if (isDefined == "XYO_PLATFORM_64BIT") {
			return VariableBoolean::newVariable(true);
		};
#endif
#ifdef XYO_PLATFORM_32BIT
		if (isDefined == "XYO_PLATFORM_32BIT") {
			return VariableBoolean::newVariable(true);
		};
#endif

// -- Thread support
#ifdef XYO_PLATFORM_SINGLE_THREAD
		if (isDefined == "XYO_PLATFORM_SINGLE_THREAD") {
			return VariableBoolean::newVariable(true);
		};
#endif
#ifdef XYO_PLATFORM_MULTI_THREAD
		if (isDefined == "XYO_PLATFORM_MULTI_THREAD") {
			return VariableBoolean::newVariable(true);
		};
#endif
		return VariableBoolean::newVariable(false);
	};

	TPointer<Variable> xyoGetPlatform(VariableFunction *function, Variable *this_, VariableArray *arguments) {
		return VariableString::newVariable(XYO_PLATFORM_NAME);
	};

	TPointer<Variable> osIsWindows(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#if defined(XYO_PLATFORM_OS_WINDOWS) || defined(XYO_PLATFORM_OS_MINGW)
		return VariableBoolean::newVariable(true);
#else
		return VariableBoolean::newVariable(false);
#endif
	};

	TPointer<Variable> osIsLinux(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#if defined(XYO_PLATFORM_OS_LINUX)
		return VariableBoolean::newVariable(true);
#else
		return VariableBoolean::newVariable(false);
#endif
	};

	TPointer<Variable> osIsMinGW(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#if defined(XYO_PLATFORM_OS_MINGW)
		return VariableBoolean::newVariable(true);
#else
		return VariableBoolean::newVariable(false);
#endif
	};

	static TPointer<Variable> getVersion(VariableFunction *function, Variable *this_, VariableArray *arguments) {
#ifdef XYO_QUANTUMSCRIPT_DEBUG_RUNTIME
		printf("- fabricare-get-version\n");
#endif
		return VariableString::newVariable(XYO::Fabricare::Version::versionWithBuild());
	};

#include <XYO/Fabricare/Library.Source.cpp>
#include <XYO/Fabricare/Internal.Source.cpp>

	void initExecutive(Executive *executive) {
		Extension::Console::registerInternalExtension(executive);
		Extension::Buffer::registerInternalExtension(executive);
		Extension::DateTime::registerInternalExtension(executive);
		Extension::JSON::registerInternalExtension(executive);
		Extension::Application::registerInternalExtension(executive);
		Extension::Shell::registerInternalExtension(executive);
		Extension::ShellFind::registerInternalExtension(executive);
		Extension::Math::registerInternalExtension(executive);
		Extension::ProcessInteractive::registerInternalExtension(executive);
		Extension::SHA512::registerInternalExtension(executive);
		Extension::Thread::registerInternalExtension(executive);
		Extension::Task::registerInternalExtension(executive);
		Extension::Job::registerInternalExtension(executive);
		Extension::Make::registerInternalExtension(executive);
		Extension::CSV::registerInternalExtension(executive);
		Extension::URL::registerInternalExtension(executive);

		executive->compileString("Script.requireExtension=Script.requireInternalExtension;");

		executive->compileStringX("var Internal={};");
		executive->setFunction2("Internal.fileToCS", internalFileToCS);
		executive->setFunction2("Internal.fileToRC", internalFileToRC);
		executive->setFunction2("Internal.htmlToRC", internalHTMLToRC);
		executive->setFunction2("Internal.xyoCC", internalXYOCC);
		executive->setFunction2("Internal.xyoVersion", internalXYOVersion);

		executive->compileStringX("var XYO={};");
		executive->setFunction2("XYO.isConfigDefined(def);", xyoIsConfigDefined);
		executive->setFunction2("XYO.getPlatform", xyoGetPlatform);

		executive->compileStringX("var OS={};");
		executive->setFunction2("OS.isWindows", osIsWindows);
		executive->setFunction2("OS.isLinux", osIsLinux);
		executive->setFunction2("OS.isMinGW", osIsMinGW);

		internalInitExecutive(executive);

		executive->compileStringX("var Fabricare={};");
		executive->setFunction2("Fabricare.getVersion()", getVersion);

		executive->compileStringX(librarySource);
	};

};
